src/Elmish.WPF/WpfProgram.fs

```diff
@@ -0,0 +1,133 @@
+﻿namespace Elmish.WPF
+
+open System.Windows
+open Microsoft.Extensions.Logging
+open Microsoft.Extensions.Logging.Abstractions
+open Elmish
+
+
+type WpfProgram<'model, 'msg> =
+  internal {
+    ElmishProgram: Program<unit, 'model, 'msg, Binding<'model, 'msg> list>
+    LoggerFactory: ILoggerFactory
+    /// Only log calls that take at least this many milliseconds. Default 1.
+    PerformanceLogThreshold: int
+  }
+
+
+[<RequireQualifiedAccess>]
+module WpfProgram =
+
+
+  let private fromElmishProgram program =
+    { ElmishProgram = program
+      LoggerFactory = NullLoggerFactory.Instance
+      PerformanceLogThreshold = 1 }
+
+
+  /// Creates a WpfProgram that does not use commands.
+  let mkSimple
+      (init: unit -> 'model)
+      (update: 'msg  -> 'model -> 'model)
+      (bindings: unit -> Binding<'model, 'msg> list) =
+    Program.mkSimple init update (fun _ _ -> bindings ())
+    |> fromElmishProgram
+
+
+  /// Creates a WpfProgram that uses commands
+  let mkProgram
+      (init: unit -> 'model * Cmd<'msg>)
+      (update: 'msg  -> 'model -> 'model * Cmd<'msg>)
+      (bindings: unit -> Binding<'model, 'msg> list) =
+    Program.mkProgram init update (fun _ _ -> bindings ())
+    |> fromElmishProgram
+
+
+  /// Starts the Elmish dispatch loop, setting the bindings as the DataContext
+  /// for the specified FrameworkElement. Non-blocking. This is a low-level function;
+  /// for normal usage, see runWindow and runWindowWithConfig.
+  let startElmishLoop
+      (element: FrameworkElement)
+      (program:  WpfProgram<'model, 'msg>) =
+    let mutable lastModel = None
+
+    let setState model dispatch =
+      match lastModel with
+      | None ->
+          let bindings = Program.view program.ElmishProgram model dispatch
+          let vm = ViewModel<'model, 'msg>(model, dispatch, bindings, program.PerformanceLogThreshold, "main", program.LoggerFactory)
+          element.DataContext <- vm
+          lastModel <- Some vm
+      | Some vm ->
+          vm.UpdateModel model
+
+    let uiDispatch (innerDispatch: Dispatch<'msg>) : Dispatch<'msg> =
+      fun msg -> element.Dispatcher.Invoke(fun () -> innerDispatch msg)
+
+    let logMsgAndModel (msg: 'msg) (model: 'model) = 
+      program.LoggerFactory.CreateLogger("Elmish.WPF.Update").LogTrace("New message: {Message}", msg)
+      program.LoggerFactory.CreateLogger("Elmish.WPF.Update").LogTrace("Updated state:\n{Model}", model)
+
+    let logError (msg: string, ex: exn) =
+      program.LoggerFactory.CreateLogger("Elmish.WPF.Update").LogError(ex, msg)
+
+    program.ElmishProgram
+    |> Program.withTrace logMsgAndModel
```

Comment:
`logMsgAndModel` only logs at the `Trace` level.  Can you optimize this so that `Program.withTrace` is only called when `ILogger.IsEnabled(LogLevel.Trace)` is `true`?

Reply:
AFAIK there's no need for that, as most log frameworks already optimize the heck out of just that. As long as you're not doing important work and just passing arguments, that is. That's why I check `IsEnabled` for the performance logger, since if it's not enabled, we shouldn't measure.

Reply:
I think the warning against premature optimization is bases on these three things.
1. The optimized code is harder to understand.
2. The optimized code is slower than the unoptimized code.
3. The optimized code is faster than the optimized code, but the time spent optimizing wasn't worth the modest speed improvement.

I don't think any of these apply in this case.
1. The optimized code is still easy to read in this case.  (I will try to figure out how to suggest a commit through GitHub.)
2. The optimized code is strictly faster (unless the compiler optimizes the function calls away, in which case, the efficiency is the same).
3. The implementation time is minimal, we are are library, so it is hard for us to tell how much extreme some use case could get.

Reply:
I figured out how to give "commitable" change suggestions and gave the suggestion I mentioned above [here](https://github.com/elmish/Elmish.WPF/pull/255/files?file-filters%5B%5D=.fs&file-filters%5B%5D=.fsproj&hide-deleted-files=true#r467651476).

Reply:
Oh, I see. I read your suggestion incorrectly and thought you meant checking the level in `ViewModel.fs` before every (Trace) log statement. I'll make the change.

Reply:
> Oh, I see. I read your suggestion incorrectly and thought you meant checking the level in ViewModel.fs before every (Trace) log statement. I'll make the change.

I would also prefer to optimize those `Trace` log statements in `ViewModel.fs`.  Not by wrapping each on in an `if` with a predicate that gets evaluated each time (say) WPF calls `TryGetMember` but just like the suggestion above, in such a way that the predicate gets evaluated (conceptually) once for each execution of the program.

However, this change isn't very important and it shouldn't be made in this PR anyway.

Reply:
See e.g. the [Serilog docs](https://github.com/serilog/serilog/wiki/Writing-Log-Events#level-detection) which say:

> In most cases, applications should write events without checking the active logging level. Level checking is extremely cheap and the overhead of calling disabled logger methods very low.
> 
> In *rare, performance-sensitive cases*, the recommended pattern for level detection is to

I'm not at all convinced the logging we're doing in ViewModel is such a *rare, performance-sensitive case*. But if you can point to some popular logging framework where normal logging statements are *not* extremely cheap, I can certainly bend the other way.

Reply:
> In rare, performance-sensitive cases, the recommended pattern for level detection is to store the results of level detection in a field, for example:
```C#
// in some class
readonly bool _isDebug = Log.IsEnabled(LogEventLevel.Debug);

// in some method in that class
if (_isDebug) Log.Debug("Someone is stuck debugging...");
```

I never want to do that.  I never want my code guard each potential call to (say) `Log.Debug` with such an `if` statement.

Instead, I want to optimize like I suggested in the initial comment in this thread and you implemented as
```F#
|> if updateLogger.IsEnabled LogLevel.Trace then Program.withTrace logMsgAndModel else id
```

This `if` statement is only executed once each time the entire application is started.  That is what I am referring to when I said
> 2. The optimized code is strictly faster (unless the compiler optimizes the function calls away, in which case, the efficiency is the same).

We pay the cost of evaluating that `if` statement once.  In exchange, we do highly optimized behavior in each case.  In particular, we never even try to log when logging at that log level is suppressed.

The performance logging code also uses the same highly efficient optimization (and you have maintained it in this branch).
```F#
let measure name callName f =
  if not <| logPerformance.IsEnabled(LogLevel.Trace) then f
  else
    fun x -> // measure, do work, and log
```

That `if` statement is executed once per `ViewModel<...>` instance (during initialization).  It could be improved to also execute just once per execution of the entire application, but this change is also a low priority and not appropriate for this PR.

Reply:
I wish there was a name for this approach to logging.  One perspective is that it is runtime optimization related to logging.  The other is that it is a very good separation of concerns.  For example, it is very pleasing to read the `ViewModel<...>` code and _not see_ (possibly suppressed) performance logging calls in all the binding cases when (say) `get` and `set` are called.

Normally my experience is that code that is easy to read and well architectured has good performance but not the best.  That is ok though because the best performance isn't needed.

In the case of logging, my claim is that this well-architectured code also has the best performance.

Reply:
Maybe I'm tired, but I'm having a hard time following exactly what you mean. But let's not spend more time on that right now, since it's not relevant for this PR.

Reply:
Sure.  (It is probably me not clearly explaining myself though.)